stable_baselines3/common/pytree_dataclass.py

```diff
@@ -1,21 +1,282 @@
+from stable_baselines3.common.utils import zip_strict
+
+__all__ = [
+    "FrozenPyTreeDataclass",
+    "MutablePyTreeDataclass",
+    "TensorTree",
+    "tree_empty",
+    "tree_flatten",
+    "tree_index",
+    "tree_map",
+]
 
 T = TypeVar("T")
+U = TypeVar("U")
 
 SB3_NAMESPACE = "stable-baselines3"
 
+_RESERVED_NAMES = ["_PyTreeDataclassBase", "FrozenPyTreeDataclass", "MutablePyTreeDataclass"]
+
+
+# We need to inherit from `type(CustomTreeNode)` to prevent conflicts due to different-inheritance in metaclasses.
+# - For some reason just inheriting from `typing._ProtocolMeta` does not get rid of that error.
+# - Inheriting from `typing._GenericAlias` is impossible, as it's a `typing._Final` class.
+#
+# But in mypy, inheriting from a dynamic base class from `type` is not supported, so we disable type checking for this
+# line.
+class _PyTreeDataclassMeta(type(CustomTreeNode)):  # type: ignore[misc]
+    """Metaclass to register dataclasses as PyTrees.
+
+    Usage:
+      class MyDataclass(metaclass=_DataclassPyTreeMeta):
+        ...
+    """
+
+    # We need to have this `currently_registering` variable because, in the course of making a DataClass with __slots__,
+    # another class is created. So this will be called *twice* for every dataclass we annotate with this metaclass.
+    currently_registering: ClassVar[Optional[type]] = None
+
+    def __new__(mcs, name, bases, namespace, slots=True, **kwargs):
+        # First: create the class in the normal way.
+        cls = super().__new__(mcs, name, bases, namespace)
+
+        if dataclasses.is_dataclass(cls):
+            # If the class we're registering is already a Dataclass, it means it is a descendant of FrozenPyTreeDataclass or
+            # MutablePyTreeDataclass.
+            # This includes the children which are created when we create a dataclass with __slots__.
+
+            if mcs.currently_registering is not None:
+                # We've already created and annotated a class without __slots__, now we create the one with __slots__
+                # that will actually get returned after from the __new__ method.
+                assert mcs.currently_registering.__module__ == cls.__module__
+                assert mcs.currently_registering.__name__ == cls.__name__
+                mcs.currently_registering = None
+                return cls
+
+            else:
+                assert name not in _RESERVED_NAMES, (
+                    f"Class with name {name}: classes {_RESERVED_NAMES} don't inherit from a dataclass, so they should "
+                    "not be in this branch."
+                )
+
+                # Otherwise we just mark the current class as what we're registering.
+                if not issubclass(cls, (FrozenPyTreeDataclass, MutablePyTreeDataclass)):
+                    raise TypeError(f"Dataclass {cls} should inherit from FrozenPyTreeDataclass or MutablePyTreeDataclass")
+                mcs.currently_registering = cls
```

Comment:
I think 87-89 can be updated to just:
`mcs.currently_registering = cls`

Like, if we enter the `if mcs.currently_registering is not None` on line 70, we will never extend past here. And if we don't enter that `if`, then we always do `mcs.currently_registering = cls`.

Reply:
You're right, thank you! This code is so confusing >.<